lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+            }
+
+            self.slots[index] = @ptrCast(*c_void, aligned_ptr);
+        }
+
+        return self.slots[index].?;
+    }
+};
+
+// global stucture for ThreadStorage.
+// it provides thread-safety for on-demand storage of Thread Objects.
+const ThreadStorage = struct {
+    var key: std.c.pthread_key_t = undefined;
+    var init_once = std.once(ThreadStorage.init);
+
+    /// Return a per thread ObjectArray with at least the expected index.
+    pub inline fn get_array(allocator: *Allocator, index: usize) !*ObjectArray {
+        if (ThreadStorage.getspecific()) |array| {
+            // ensure the array is big enough for the wanted index
+            try array.ensureSize(index);
+            return array;
+        } else {
+            // a new array should contains at least 16 objects.
+            const size = std.math.max(16, index);
+
+            // create a new array and store it.
+            var array: *ObjectArray = try ObjectArray.initAlloc(allocator, size);
+            ThreadStorage.setspecific(array);
+            return array;
+        }
+    }
+
+    /// Return casted thread specific value.
+    inline fn getspecific() ?*ObjectArray {
+        return @ptrCast(
+            ?*ObjectArray,
+            @alignCast(
+                @alignOf(?*ObjectArray),
+                std.c.pthread_getspecific(ThreadStorage.key),
+            ),
+        );
+    }
+
+    /// Set casted thread specific value.
+    inline fn setspecific(new: ?*ObjectArray) void {
+        if (std.c.pthread_setspecific(ThreadStorage.key, @ptrCast(*c_void, new)) != 0)
+            @panic("pthread_setspecific");
+    }
+
+    /// Initialize pthread_key_t.
+    fn init() void {
+        if (std.c.pthread_key_create(&ThreadStorage.key, ThreadStorage.deinit) != 0)
+            @panic("pthread_key_create");
+    }
+
+    /// Invoked by pthread specific destructor. the passed argument is the ObjectArray pointer.
+    fn deinit(arrayPtr: *c_void) callconv(.C) void {
+        var array = @ptrCast(
+            *ObjectArray,
+            @alignCast(@alignOf(*ObjectArray), arrayPtr),
+        );
+        array.deinit();
+    }
+};
+
+// global counter for keeping track of requested indexes.
+const Index = struct {
+    var next: usize = 1;
+    var mutex: std.c.pthread_mutex_t = std.c.PTHREAD_MUTEX_INITIALIZER;
+
+    /// Return a next free index, incrementing global counter.
+    pub fn new() usize {
+        // assert(Index.mutex held);
+        const index = Index.next;
+        Index.next += 1;
+
+        return index;
+    }
+
+    pub fn lock() void {
+        if (std.c.pthread_mutex_lock(&Index.mutex) != 0) {
+            @panic("pthread_mutex_lock");
+        }
+    }
+
+    pub fn unlock() void {
+        if (std.c.pthread_mutex_unlock(&Index.mutex) != 0) {
+            @panic("pthread_mutex_unlock");
+        }
+    }
+};
+
+const gcc_word = usize; // XXX typedef unsigned int gcc_word __attribute__((mode(word)));
```

Comment:
Please keep all the type definitions at the top of the file.

This is meant to be largest unsigned int that can fit in a register so for all we care (right now) `usize` is ok.